### PR TITLE
vmss managed identity

### PR DESCRIPTION
**Pull request by abdullahaslam306**

### CODE REVIEW

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
+var async = require('async');
+
```
Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js:2]:
```suggestion
```

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+var helpers = require('../../../helpers/azure');
+
```
Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js:4]:
```suggestion
```

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+module.exports = {
+    title: 'VM Scale Set Managed Identity Enabled',
+    category: 'Virtual Machines',
+    domain: 'Compute',
+    description: 'Ensures that Azure Virtual Machine scale sets have managed identity enabled.',
```
Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js:10]:
```suggestion
    description: 'Ensures that Virtual Machine Scale Sets have managed identity enabled.',
```

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
+var async = require('async');
+var helpers = require('../../../helpers/azure');
+
+module.exports = {
+    title: 'VM Scale Set Managed Identity Enabled',
+    category: 'Virtual Machines',
+    domain: 'Compute',
+    description: 'Ensures that Azure Virtual Machine scale sets have managed identity enabled.',
```
Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js:8]:
```suggestion
    description: 'Ensures that Azure Virtual Machine Scale Sets have managed identity enabled.',
```

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    more_info: 'Managed identities for Azure resources provide Azure services with an automatically managed identity in Microsoft Entra ID. You can use this identity to authenticate to any service that supports Microsoft Entra authentication, without having credentials in your code.',
```
Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js:9]: can we make it consistent like we have in other managed identity plugins ,
Enabling managed identities eliminate the need for developers having to manage credentials by providing an identity for the Azure resource in Azure AD and using it to obtain Azure Active Directory (Azure AD) tokens.

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    link: 'https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/qs-configure-portal-windows-vmss',
+    recommended_action: 'Modify VM scale set and enable user or system assigned identities.',
```
Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js:11]:
```suggestion
    recommended_action: 'Modify VM Scale Set and enable managed identity.',
```

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+var vmssManagedIdentityEnabled = require('./vmssManagedIdentityEnabled');
+
+const virtualMachineScaleSets = [
+    {
+        'name': 'test-vmss',
+        'id': '/subscriptions/123/resourceGroups/AQUA-RESOURCE-GROUP/providers/Microsoft.Compute/virtualMachineScaleSets/test-vmss',
+        'type': 'Microsoft.Compute/virtualMachineScaleSets',
+        "identity": {
+            "type": "SystemAssigned",
+            "principalId": "5db3ed52-909e-4016-a31f-d2fe043952a4",
```
Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js:11]: use dummy values here as well for principalId

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('VM scale set has managed identity enabled');
+                expect(results[0].region).to.equal('eastus');
+                done();
+            });
+        });
+
+        it('should give failing result if VM scale set does not have managed identity enabled', function(done) {
+            const cache = createCache([virtualMachineScaleSets[1]]);
+            vmssManagedIdentityEnabled.run(cache, { vmss_approved_extensions: 'healthRepairExtension' }, (err, results) => {
```
Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js:73]: is this setting? { vmss_approved_extensions: 'healthRepairExtension' }

**Review by alphadev4 — CHANGES_REQUESTED** [review]

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js (hunk 8 above)
Comment (abdullahaslam306) [reply plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js:73]: yes

Comment (alphadev4) [reply plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js:73]: but we don't have setting in this plugin

Comment (abdullahaslam306) [reply plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js]: my bad resolved

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+            }
+            if (!vmScaleSets.data.length) {
+                helpers.addResult(results, 0, 'No existing Virtual Machine Scale Sets found', location);
+                return rcb();
+            }
+            for (let scaleSet of vmScaleSets.data) {
+                if (!scaleSet.id) continue;
+
+                if (scaleSet.identity && scaleSet.identity.type){
+                    helpers.addResult(results, 0, 'VM scale set has managed identity enabled', location, scaleSet.id);
```
Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js:37]:
```suggestion
                    helpers.addResult(results, 0, 'Virtual Machine Scale Set has managed identity enabled', location, scaleSet.id);
```

Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js:37]:
```suggestion
                    helpers.addResult(results, 0, 'Virtual Machine Scale Set has managed identity enabled', location, scaleSet.id);
```

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
+var async = require('async');
+var helpers = require('../../../helpers/azure');
+
+module.exports = {
+    title: 'VM Scale Set Managed Identity Enabled',
+    category: 'Virtual Machine Scale Set',
+    domain: 'Compute',
+    description: 'Ensures that Azure Virtual Machine Scale Sets have managed identity enabled.',
```
Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js:8]:
```suggestion
    description: 'Ensures that Azure Virtual Machine Scale Set have managed identity enabled.',
```

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+                helpers.addResult(results, 0, 'No existing Virtual Machine Scale Sets found', location);
+                return rcb();
+            }
+            for (let scaleSet of vmScaleSets.data) {
+                if (!scaleSet.id) continue;
+
+                if (scaleSet.identity && scaleSet.identity.type){
+                    helpers.addResult(results, 0, 'VM scale set has managed identity enabled', location, scaleSet.id);
+                } else {
+                    helpers.addResult(results, 2, 'VM scale set does not have managed identity enabled', location, scaleSet.id);
```
Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.js:39]:
```suggestion
                    helpers.addResult(results, 2, 'Virtual Machine Scale Set does not have managed identity enabled', location, scaleSet.id);
```

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+                done();
+            });
+        });
+
+        it('should give passing result if VM scale set has managed identity enabled', function(done) {
+            const cache = createCache([virtualMachineScaleSets[0]]);
+            vmssManagedIdentityEnabled.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('VM scale set has managed identity enabled');
```
Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js:65]:
```suggestion
                expect(results[0].message).to.include('Virtual Machine Scale Set has managed identity enabled');
```

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+                done();
+            });
+        });
+
+        it('should give failing result if VM scale set does not have managed identity enabled', function(done) {
+            const cache = createCache([virtualMachineScaleSets[1]]);
+            vmssManagedIdentityEnabled.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                expect(results[0].message).to.include('VM scale set does not have managed identity enabled');
```
Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js:76]:
```suggestion
                expect(results[0].message).to.include('Virtual Machine Scale Set does not have managed identity enabled');
```

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js (hunk 7 above)
Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js:11]:
```suggestion
            "principalId": "123454",
```

File: plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+const virtualMachineScaleSets = [
+    {
+        'name': 'test-vmss',
+        'id': '/subscriptions/123/resourceGroups/AQUA-RESOURCE-GROUP/providers/Microsoft.Compute/virtualMachineScaleSets/test-vmss',
+        'type': 'Microsoft.Compute/virtualMachineScaleSets',
+        "identity": {
+            "type": "SystemAssigned",
+            "principalId": "5db3ed52-909e-4016-a31f-d2fe043952a4",
+            "tenantId": "d207c7bd-fcb1-4dd3-855a-cfd2f9b651e8"
```
Comment (alphadev4) [inline plugins/azure/virtualmachinescaleset/vmssManagedIdentityEnabled.spec.js:12]:
```suggestion
            "tenantId": "1234548"
```

**Review by alphadev4 — APPROVED** [review]

**Review by alphadev4 — APPROVED** [review]

**Pull request merged**